File: ITP_KANDY_7-main/backend/src/domain/entities/Customer.js
/**
 * Domain Layer: Customer Entity.
 * Represents a consumer or person of interest in the merchant's CRM.
 * Encapsulates contact details, creditworthiness (Limits), and current debt (Outstanding).
 */
class Customer {
    /**
     * Logic: Entity Initialization.
     * Maps consumer profile data into a structured business object for credit management.
     */
    constructor({ 
        id, // Unique identity (UUID)
        name, // Consumer full name or alias
        phone, // Primary identification and communication key
        imageUrl, // Optional profile avatar link
        totalOutstanding, // Aggregate debt: The amount of money this customer currently owes the shop
        creditLimit, // Policy: The maximum amount of debt allowed for this account
        status, // State: 'active' or 'suspended' (blocks credit sales if suspended)
        lastPurchase, // Date of the most recent transactional activity
        createdAt, // Audit: Genesis date
        updatedAt // Audit: Last profile modification timestamp
    }) {
        this.id = id;
        this.name = name || '';
        this.phone = phone || '';
        this.imageUrl = imageUrl || '';
        this.totalOutstanding = totalOutstanding || 0; // Logic: Incremented on credit sales, decremented on payments
        this.creditLimit = creditLimit || 0;
        this.status = status || 'active';
        this.lastPurchase = lastPurchase || '';
        this.createdAt = createdAt || new Date().toISOString(); // Stability: Fallback to current time if missing
        this.updatedAt = updatedAt || new Date().toISOString();
    }

    /**
     * Logic: Data Serialization.
     * Converts the entity into a JSON object for API transmission to the mobile app.
     */
    toJSON() {
        return {
            id: this.id,
            name: this.name,
            phone: this.phone,
            imageUrl: this.imageUrl,
            totalOutstanding: this.totalOutstanding, // Critical: Used for 'Over Limit' validation logic
            creditLimit: this.creditLimit,
            status: this.status,
            lastPurchase: this.lastPurchase,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt,
        };
    }
}

// Module Export: Business logic representation of a customer/debtor.
module.exports = Customer;

File: ITP_KANDY_7-main/frontend/lib/features/credit/domain/entities/customer.dart
// ------------------------------------------------------------------------------
// File: customer.dart
// Purpose: Domain entity representing a credit-tracked customer.
// Rationale: Provides a strongly-typed model for customer profile data
//   including outstanding debt balance, used across Credit module screens,
//   providers, and PDF export utilities.
// ------------------------------------------------------------------------------
class Customer {
  final String id;
  final String name;
  final String phone;
  final String imageUrl;
  final double totalOutstanding;
  final double creditLimit;
  final String status;
  final String lastPurchase;
  final String createdAt;

  Customer({
    required this.id,
    required this.name,
    this.phone = '',
    this.imageUrl = '',
    this.totalOutstanding = 0, // The current amount of money this customer owes the shop
    this.creditLimit = 5000, // The maximum debt this customer is allowed to incur
    this.status = 'active',
    this.lastPurchase = '',
    this.createdAt = '',
  });

  factory Customer.fromJson(Map<String, dynamic> json) {
    return Customer(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      phone: json['phone'] ?? '',
      imageUrl: json['imageUrl'] ?? '',
      totalOutstanding:
          double.tryParse(json['totalOutstanding']?.toString() ?? '0') ?? 0.0,
      creditLimit:
          double.tryParse(json['creditLimit']?.toString() ?? '0') ?? 0.0,
      status: json['status'] ?? 'active',
      lastPurchase: json['lastPurchase'] ?? '',
      createdAt: json['createdAt'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {'name': name, 'phone': phone, 'creditLimit': creditLimit};
  }
}

